### DISCUSSION TITLE
Shared vs independent contributions

### DISCUSSION THREAD

**Discussion by MarcRieraDominguez** [jluchman/domir#9] [General]
Good afternoon!
I would like to know more about how the results of dominance analysis relate to "joint" and "independent" contributions to explained variation. I sometimes read analyses that report a model Y ~ A + B, indicating variance explained soley by A, solely by B, and jointly by A and B. How does dominance analysis relate to such approaches to variation partitioning?
Apologies if the language is not very technical!
Thank you!
Marc

**Comment by fmg-jluchman** [comment]
Hi Marc,

  I believe you're referring to 'commonality analysis'.  There's an article in _Organizational Research Methods_ which discusses these ideas, and others, compared to one another (linked below). 

https://journals.sagepub.com/doi/abs/10.1177/1094428113493929

  In short, commonality analysis makes a distinction between shared and unique values where domiance analysis averages over those shared values to ascribe them to variables directly.

  For instance, using the _mtcars_ data, the uniqueness of _cyl_ and _am_ are 0.3992 and 0.0328, respectively. Their commonality is then .3628.  

 The dominance analysis uses their uniquenesses (equal to their conditional dominance statistics with 2 IVs) but also their unadjusted squared correlation. these values are averaged to get the general dominance statistic.

  The general dominance statistics ascribed to each of the IVs then partition up the .3628 commonality value which you can see when looking at their values.  _cyl_ resulted in .5627 and _am_ obtained .1963. Literally, ascribed half of the commonality of .3628 to both (~.1635 added to each uniqueness). General dominance then just acknowledges that in situations like this, best guess is to split the difference and ascribe half to each.  It gets more complex with more IVs and suppression effects but that's the general idea.  

- joe

**Reply by MarcRieraDominguez** [reply]
Dear Joe,
Thank you very much for your reply, very illuminating!
I'll be closing this.
Cheers,
Marc

**Comment by MarcRieraDominguez** [comment]
Hi Joe,
I would like to ask a follow up question. I apologize in advance as it lies a bit outside the scope of dominance analysis, although it falls within the scope of the relative importance of predictors in multiple regression.
I would like to run a commonality analysis on a binomial GLM, but I can't seem to find a suitable package to this. Given that I have six-something predictors, I verymuch favor an authomated procedure. I have checked `yhat::CommonaliyCoefficients()`, but it seems to fit only `lm`. Could you advice on a suitable package to fit a Commonality analysis to a binomial GLM, in R?
Many thanks in advance!
Marc

**Comment by jluchman** [comment]
Good question here and I am not familiar with other packages that decompose fit statistics with a commonality focus other than the _yhat_ package.

This is something that I have considered building into _domir_ since your question previously, but there are a few other development priorities to implement before I would be able to turn to it.

With a bit of work, you might be able to 'trick' `yhat::commonality` into giving you the right coefficients. It takes input from `yhat::aps` and if you were to run this with with our data 'as though' it were an `lm` then loop through the values in the _APSMatrix_ it returns and replace them with the pseudo-R2 value generated by the model when it is estimated by `glm`, I suspect the underlying method will return the correct commonality coefficients for the `glm` with the desired pseudo-R2.  That is likely the easiest way to implement in the short term.

**Reply by fmg-jluchman** [reply]
As an example of the workaround:

```
> cc1 <- cc2 <- aps(mtcars, "vs", c("mpg", "wt", "drat"))
> commonality(cc1)
             Coefficient      % Total
mpg         1.307474e-01 2.952882e-01
wt          1.560625e-03 3.524612e-03
drat        5.445407e-06 1.229825e-05
mpg,wt      1.166259e-01 2.633952e-01
mpg,drat    4.094771e-03 9.247888e-03
wt,drat     2.652812e-04 5.991277e-04
mpg,wt,drat 1.894796e-01 4.279327e-01
> for (ivset in 1:nrow(cc1$APSMatrix)) {
+  parse_iv <- strsplit(rownames(cc1$APSMatrix)[[ivset]], ",")[[1]]
+  print(new_r2 <- 
+     reformulate(parse_iv, response = "vs") |> 
+     glm(family = binomial(), data = mtcars) |> 
+     performance::r2() |> _[[1]])
+  cc2$APSMatrix[ivset,2] <- new_r2
+}
Tjur's R2 
0.4738108 
Tjur's R2 
0.3276353 
Tjur's R2 
0.2026548 
Tjur's R2 
0.4776926 
Tjur's R2 
0.4745717 
Tjur's R2 
0.3275208 
Tjur's R2 
0.4770796 
> commonality(cc2)
              Coefficient      % Total
mpg          0.1495588050  0.313488152
wt           0.0025079168  0.005256810
drat        -0.0006130118 -0.001284926
mpg,wt       0.1223580938  0.256473116
mpg,drat     0.0004985007  0.001044901
wt,drat      0.0013739076  0.002879829
mpg,wt,drat  0.2013953974  0.422142119
```

**Reply by MarcRieraDominguez** [reply]
Many thanks Joe for looking into this! As far as I can tell it makes sense, but I'm unable to reproduce the results above. Could you eaborate on the `ivset` object sinde the loop and the matrix subsetting, and what structure should it have? Thank you!!

**Reply by jluchman** [reply]
The rownames of the `.$APSMatrix` object returned by `aps()` contain all predictor combinations and the code uses that feature to update this matrix.  It's structure as is below.

```
cc1$APSMatrix

            k        R2
mpg         1 0.4409477
wt          1 0.3079314
drat        1 0.1938451
mpg,wt      2 0.4427736
mpg,drat    2 0.4412184
wt,drat     2 0.3120316
mpg,wt,drat 3 0.4427790
```

The `commonality()` function uses that matrix directly but, in it's present state, the $R^2$ values are based on `lm()` and not `glm()` with a pseudo-$R^2$. `ivset` is an integer number intended to select one element of the rownames of the `.$APSMatrix`, parses them into a vector with `strsplit()` then `reformulates()` and feeds them into a pipeline that estimates the pseudo-$R^2$ and records it in the correct location in a version of the `.$APSMatrix`.  

Updated the function a bit to show what it uses in the `reformulate()` call.

```
for (ivset in 1:nrow(cc1$APSMatrix)) {
  parse_iv <- strsplit(rownames(cc1$APSMatrix)[[ivset]], ",")[[1]]
  print(parse_iv)
  new_r2 <- 
    reformulate(parse_iv, response = "vs") |> 
    glm(family = binomial(), data = mtcars) |> 
    performance::r2() |>
    _[[1]] |>
    print()
  cc2$APSMatrix[ivset,2] <- new_r2
}

[1] "mpg"
Tjur's R2 
0.4738108 
[1] "wt"
Tjur's R2 
0.3276353 
[1] "drat"
Tjur's R2 
0.2026548 
[1] "mpg" "wt" 
Tjur's R2 
0.4776926 
[1] "mpg"  "drat"
Tjur's R2 
0.4745717 
[1] "wt"   "drat"
Tjur's R2 
0.3275208 
[1] "mpg"  "wt"   "drat"
Tjur's R2 
0.4770796 
```

**Reply by MarcRieraDominguez** [reply]
Many thanks, all clear now! :)
